man/man8/xfs_estimate.8
.TH xfs_estimate 8
.SH NAME
xfs_estimate \- estimate the space that an XFS filesystem will take
.SH SYNOPSIS
.nf
\f3xfs_estimate\f1 [ \f3\-h?\f1 ] [ \f3\-b\f1 blocksize ] [ \f3\-i\f1 logsize ]
		   [ \f3\-e\f1 logsize ] [ \f3\-v\f1 ] directory ...
.fi
.SH DESCRIPTION
For each \f2directory\f1 argument,
.I xfs_estimate
estimates the space that directory would take if it were copied to an XFS
filesystem.
.I xfs_estimate
does not cross mount points.
The following definitions
are used:
.PD 0
.IP
KB = *1024
.IP
MB = *1024*1024
.IP
GB = *1024*1024*1024
.PD
.PP
The
.I xfs_estimate
options are:
.TP
\f3\-b\f1 \f2blocksize\f1
Use
.I blocksize
instead of the default blocksize of 4096 bytes.
The modifier
.B k
can be used
after the number to indicate multiplication by 1024.
For example,
.sp .8v
.RS
	\f4xfs_estimate -b 64k /\f1
.RE
.IP
requests an estimate of the space required by the directory / on an
XFS filesystem using a blocksize of 64K (65536) bytes.
.TP
.B \-v
Display more information, formatted.
.TP
.B \-h
Display usage message.
.TP
.B \-?
Display usage message.
.TP
\f3\-i, \-e\f1 \f2logsize\f1
Use
.I logsize
instead of the default log size of 1000 blocks.
.B \-i
refers to an internal log, while
.B \-e
refers to an external log.
The modifiers
.B k
or
.B m
can be used
after the number to indicate multiplication by 1024 or 1048576, respectively.
.IP
For example,
.sp .8v
.RS
	\f4xfs_estimate -i 1m /\f1
.RE
.IP
requests an estimate of the space required by the directory / on an
XFS filesystem using an internal log of 1 megabyte.
.SH EXAMPLES
.nf
.sp 8v
% \f4xfs_estimate -e 10m /var/tmp\f1\f7
/var/tmp will take about 4.2 megabytes
        with the external log using 2560 blocks or about 10.0 megabytes
.fi
.nf
.sp .8v
% \f4xfs_estimate -v -e 10m /var/tmp\f1\f7
directory                     bsize   blocks    megabytes    logsize
/var/tmp                       4096      792        4.0MB   10485760
.fi
.nf
.sp .8v
% \f4xfs_estimate -v /var/tmp\f1\f7
directory                     bsize   blocks    megabytes    logsize
/var/tmp                       4096     3352       14.0MB   10485760
.fi
.nf
.sp .8v
% \f4xfs_estimate /var/tmp\f1\f7
/var/tmp will take about 14.0 megabytes
.fi
.Ee


doc/CHANGES
xfsdump-2.2.21 (15 Apr 2004)
	- Always allocate a filesystem handle on restore instead of
	  doing so only when restoring DMAPI events, as the filesystem
	  handle is also needed if restoring extended attrs.
	- Remove path_to_fshandle calls in tree.c. These existed to
	  ensure that we always had a fshandle before calling
	  open_by_handle. These calls are extraneous since we now
	  allocate a fshandle during context_init.
	- Change open_by_fshandle calls to open_by_handle since
	  we are passing a file handle, not a fshandle.

xfsdump-2.2.20 (31 Mar 2004)
	- fix up "xfs_invutil -i" which was getting segmentation faults on
	  presumably long session labels.
	  See http://oss.sgi.com/bugzilla/show_bug.cgi?id=320.
	  Become paranoid and convert all sprintf's to snprintf's.

xfsdump-2.2.19 (23 Mar 2004)
	- rearrange fsr algorithm to not copy data unless extent
	  count has been improved.  Suggested by Chris Wedgwood
	  <cw at f00f dot org> and Utz Lehmann <xfs at s2y4n2c dot de>

xfsdump-2.2.18 (25 Feb 2004)
	- add support for the security extended attributes namespace

xfsdump-2.2.17 (13 Feb 2004)
	- modify calls to libhandle subroutines to account for
	  changes to the libhandle library.

xfsdump-2.2.16 (05 Dec 2003)
	- fix xfsrestore to restore new XFS inode flags:  immutable,
	  append, sync, noatime and nodump.  The inode flags have also
	  been added to the _mk_fillconfig_ea() routine in common.dump,
	  called by QA test 063 to test inode flags and xattrs.
	- fix xfsdump to honor the new nodump xflag, "XFS_XFLAG_NODUMP",
	  (in conjunction with -e).  The original method to skip files,
	  -e plus extended attribute "SGI_XFSDUMP_SKIP_FILE", is still
	  supported.  Admins may choose which method to use.  The
	  xfsdump man page has been updated to document these changes.
	- update man pages with a more detailed description of the
	  inventory media file that gets written to the end of each dump.
	- remove 'miniroot' reference from xfsdump man page.

xfsdump-2.2.15 (18 Dec 2003)
	- a mod was added in 2.2.13 to fix the way xfsdump handles
	  multiple dump sessions to a single tape.  The mod fixed
	  problems that the TS tape driver was hitting but actually
	  broke multiple backup functionality using the ST tape
	  driver.  Both problems (ST and TS) are fixed in 2.2.15.

xfsdump-2.2.14 (18 October 2003)
	- update Debian packaging to require latest dmapi packages.

xfsdump-2.2.13 (21 July 2003)
	- fix ST/TS tape driver compatibility issues in drive_scsitape.
	- fix multiple backups to a single tape (TS AND ST).

xfsdump-2.2.12 (29 May 2003)
	- fix up an invutil Makefile botch which was effectively
	  disabling the optional ncurses support (interactive mode).

xfsdump-2.2.11 (26 April 2003)
	- rework configure.in to make use of shared macros.
	- fix up #include lines to allow compilation of sources
	  with installed headers rather than local headers.
	- update Debian packaging.
	- fix up a botch in the RPM packaging from 2.2.8.

xfsdump-2.2.10 (14 April 2003)
	- fix configure tests that used AC_PATH_PROG incorrectly.
	- switch from using mktemp to using mkstemp for xfs_copy log.
	- use a FHS compliant name for the xfs_copy log file.

xfsdump-2.2.9 (27 March 2003)
	- add -f option to xfsdq(8) to specify an output file instead
	  of using the standard output stream.  This file is created
	  by xfsdq and xfsdq will fail to run if it exists already.
	  The file is also created with a more appropriate mode than
	  whatever the umask happened to be when xfsdump(8) was run.

xfsdump-2.2.8 (27 March 2003)
	- cleanups to the build process, in particular the automatic
	  rpm package generation Makefile and spec file template.
	- Makepkgs script can now propagate failures back from make.

xfsdump-2.2.7 (13 March 2003)
	- add initial support for TS tape driver.
	  (19 February 2003)
	- fix xfsdump -I option to set correct fsid.

xfsdump-2.2.6 (19 December 2002)
	- add initial internationalisation support.

xfsdump-2.2.5 (19 December 2002)
	- fix build fallout from macro changes in XFS headers.
	- add a configure check so that we don't attempt to build
	  against old versions of the headers.

xfsdump-2.2.4 (08 November 2002)
	- configure now looks for, and xfsinvutil uses, the ncurses
	  header and library rather than the curses version - this
	  resolves an issue in those distributions that don't make
	  the compatibility library available.

xfsdump-2.2.3 (31 October 2002)
	- minor update to Debian packaging dependencies.
	- inventory related bugfix.

xfsdump-2.2.2 (03 October 2002)
	- Remove fork/setuid from fsr.  Just chown temp. file to owner.
	  This gets fsr working again with the restricted RESVSP ioctl.
	  Chris Wedgwood <cw at f00f dot org>

xfsdump-2.2.1 (24 September 2002)
	- minor troff-version-specific fixups for xfsrestore(8).

xfsdump-2.2.0 (03 September 2002)
	- force use of version 1 XFS geometry ioctl everywhere in xfsdump
	  package, so that these tools can still work on older kernels
	  (they don't need the additional information in the new ioctl).
	- added new config.h.in and configure changes to implement this.
	- do some additional cleanup after autoconf has run.

xfsdump-2.1.5 (14 August 2002)
	- minor build fix when using relatively old versions of glibc
	  which don't provide the SYS_getdents64 symbol.

xfsdump-2.1.4 (01 August 2002)
	- xfsinvutil -m option added (merge from IRIX).
	- add missing GPL/copyright headers to several xfsinvutil files.
	- correct several uses of uuid_compare in xfsinvutil.

xfsdump-2.1.3 (30 July 2002)
	- small xfsinvutil fixups to preserve compatibility with old -n
	  option.
	- fixed all new xfsinvutil warnings on 64bit platforms

xfsdump-2.1.2 (29 July 2002)
	- xfsinvutil interactive mode
	- merge of a dump/restore fix from IRIX - test if a child exited
	  with error and if a child did, set the exit_code to EXIT_ERROR.

xfsdump-2.1.1 (04 July 2002)
	- build infrastructure updates so that configure options can be

estimate/xfs_estimate.c
/*
 * Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it would be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * Further, this software is distributed without any warranty that it is
 * free of the rightful claim of any third person regarding infringement
 * or the like.  Any license provided herein, whether implied or
 * otherwise, applies only to this software file.  Patent licenses, if
 * any, provided herein do not apply to combinations of this program with
 * other software, or any other product whatsoever.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write the Free Software Foundation, Inc., 59
 * Temple Place - Suite 330, Boston MA 02111-1307, USA.
 *
 * Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
 * Mountain View, CA  94043, or:
 *
 * http://www.sgi.com
 *
 * For further information regarding this notice, see:
 *
 * http://oss.sgi.com/projects/GenInfo/SGIGPLNoticeExplan/
 */

/*
 * Estimate space of an XFS filesystem
 */
#include <libxfs.h>
#include <sys/stat.h>
#include <ftw.h>

unsigned long long
cvtnum(char *s)
{
	unsigned long long i;
	char *sp;

	i = strtoll(s, &sp, 0);
	if (i == 0 && sp == s)
		return 0LL;
	if (*sp == '\0')
		return i;
	if (*sp =='k' && sp[1] == '\0')
		return 1024LL * i;
	if (*sp =='m' && sp[1] == '\0')
		return 1024LL * 1024LL * i;
	if (*sp =='g' && sp[1] == '\0')
		return 1024LL * 1024LL * 1024LL * i;
	return 0LL;
}

int ffn(const char *, const struct stat64 *, int, struct FTW *);

#define BLOCKSIZE	4096
#define INODESIZE	256
#define PERDIRENTRY	\
	(sizeof(xfs_dir_leaf_entry_t) + sizeof(xfs_dir_leaf_name_t))
#define LOGSIZE		1000

#define FBLOCKS(n)	((n)/blocksize)
#define RFBYTES(n)	((n) - (FBLOCKS(n) * blocksize))

unsigned long long dirsize=0;		/* bytes */
unsigned long long logsize=LOGSIZE*BLOCKSIZE;	/* bytes */
unsigned long long fullblocks=0;	/* FS blocks */
unsigned long long isize=0;		/* inodes bytes */
unsigned long long blocksize=BLOCKSIZE;
unsigned long long nslinks=0;		/* number of symbolic links */
unsigned long long nfiles=0;		/* number of regular files */
unsigned long long ndirs=0;		/* number of directories */
unsigned long long nspecial=0;		/* number of special files */
unsigned long long verbose=0;		/* verbose mode TRUE/FALSE */

int __debug = 0;
int ilog = 0;
int elog = 0;

void
usage(char *progname)
{
	fprintf(stderr, "Usage: %s [opts] directory [directory ...]\n\
\t-b blocksize (fundamental filesystem blocksize)\n\
\t-i logsize (internal log size)\n\
\t-e logsize (external log size)\n\
\t-v prints more verbose messages\n\
\t-h prints this usage message\n\n\
Note:\tblocksize may have 'k' appended to indicate x1024\n\
\tlogsize may also have 'm' appended to indicate (1024 x 1024)\n",
		basename(progname));
	exit(1);
}

int
main(int argc, char **argv)
{
	unsigned long long est;
	extern int optind;
	extern char *optarg;
	char dname[40];
	int c;

	while ((c = getopt (argc, argv, "b:hdve:i:V")) != EOF) {
		switch (c) {
		case 'b':
			blocksize=cvtnum(optarg);
			if (blocksize <= 0LL) {
				fprintf(stderr, "blocksize %llu too small\n",
					blocksize);
				usage(argv[0]);
			}
			else if (blocksize > 64LL * 1024LL) {
				fprintf(stderr, "blocksize %llu too large\n",
					blocksize);
				usage(argv[0]);
			}
			break;
		case 'i':
			if (elog) {
				fprintf(stderr, "already have external log "
					"noted, can't have both\n");
				usage(argv[0]);
			}
			logsize=cvtnum(optarg);
			ilog++;
			break;
		case 'e':
			if (ilog) {
				fprintf(stderr, "already have internal log "
					"noted, can't have both\n");
				usage(argv[0]);
			}
			logsize=cvtnum(optarg);
			elog++;
			break;
		case 'v':
			verbose = 1;
			break;
		case 'd':
			__debug++;
			break;
		case 'V':
			printf("%s version %s\n", basename(argv[0]), VERSION);
			exit(0);
		default:
		case 'h':
			usage(argv[0]);
		}
	}

	if (optind == argc)
		usage(argv[0]);

	if (!elog && !ilog) {
		ilog=1;
		logsize=LOGSIZE * blocksize;
	}
	if (verbose)
		printf(
"directory                               bsize   blocks    megabytes    logsize\n");

	for ( ; optind < argc; optind++) {
		dirsize=0LL;		/* bytes */
		fullblocks=0LL;		/* FS blocks */
		isize=0LL;		/* inodes bytes */
		nslinks=0LL;		/* number of symbolic links */
		nfiles=0LL;		/* number of regular files */
		ndirs=0LL;		/* number of directories */
		nspecial=0LL;		/* number of special files */

		nftw64(argv[optind], ffn, 40, FTW_PHYS | FTW_MOUNT);

		if (__debug) {
			printf("dirsize=%llu\n", dirsize);
			printf("fullblocks=%llu\n", fullblocks);
			printf("isize=%llu\n", isize);

			printf("%llu regular files\n", nfiles);
			printf("%llu symbolic links\n", nslinks);
			printf("%llu directories\n", ndirs);
			printf("%llu special files\n", nspecial);
		}

		est = FBLOCKS(isize) + 8	/* blocks for inodes */
			+ FBLOCKS(dirsize) + 1	/* blocks for directories */
			+ fullblocks  		/* blocks for file contents */
			+ (8 * 16)	/* fudge for overhead blks (per ag) */
			+ FBLOCKS(isize / INODESIZE); /* 1 byte/inode for map */

		if (ilog)
			est += (logsize / blocksize);

		if (!verbose) {
			printf("%s will take about %.1f megabytes\n",
				argv[optind],
				(double)est*(double)blocksize/(1024.0*1024.0));
		} else {
			/* avoid running over 39 characters in field */
			strncpy(dname, argv[optind], 40);
			dname[39] = '\0';
			printf("%-39s %5llu %8llu %10.1fMB %10llu\n",
			dname, blocksize, est,
			(double)est*(double)blocksize/(1024.0*1024.0), logsize);
		}

		if (!verbose && elog) {
			printf("\twith the external log using %llu blocks ",
			logsize/blocksize);
			printf("or about %.1f megabytes\n",
			(double)logsize/(1024.0*1024.0));
		}
	}
	return 0;
}

int
ffn(const char *path, const struct stat64 *stb, int flags, struct FTW *f)
{
	/* cases are in most-encountered to least-encountered order */
	dirsize+=PERDIRENTRY+strlen(path);
	isize+=INODESIZE;
	switch (S_IFMT & stb->st_mode) {
	case S_IFREG:			/* regular files */
		fullblocks+=FBLOCKS(stb->st_blocks * 512 + blocksize-1);
		if (stb->st_blocks * 512 < stb->st_size)
			fullblocks++;	/* add one bmap block here */
		nfiles++;
		break;
	case S_IFLNK:			/* symbolic links */
		if (stb->st_size >= (INODESIZE - (sizeof(xfs_dinode_t)+4)))
			fullblocks+=FBLOCKS(stb->st_size + blocksize-1);
		nslinks++;
		break;
	case S_IFDIR:			/* directories */
		dirsize+=blocksize;	/* fudge upwards */
		if (stb->st_size >= blocksize)
			dirsize+=blocksize;
		ndirs++;
		break;
	case S_IFIFO:			/* named pipes */
	case S_IFCHR:			/* Character Special device */
	case S_IFBLK:			/* Block Special device */
	case S_IFSOCK:			/* socket */
		nspecial++;
		break;
	}
	return 0;
}
